wry/AMTKVM.py

```diff
@@ -0,0 +1,248 @@
+#                ['%s: %s' % (value.__repr__(), value in self.options) for value in self.values]
+#        )
+#
+#    def __str__(self):
+#        return self.__repr__()
+#
+#    def __iadd__(self, value):
+#        try:
+#            self.options['iadd'](value)
+#        except (KeyError, TypeError):
+#            raise AttributeError('No iadd function specified.')
+#
+#    def __isub__(self, value):
+#        try:
+#            self.options['isub'](value)
+#        except (KeyError, TypeError):
+#            raise AttributeError('No isub function specified.')
+#
+#    def toggle(self, value):
+#        if value not in self.values:
+#            raise TypeError('%r is an invalid value. Choose one of %r.' % (value, self.values))
+#        try:
+#            self._enabled_values.remove(value)
+#        except ValueError:
+#            self._enabled_values.append(value)
+#
+#    @property
+#    def enabled(self):
+#        return self._enabled_values
+#
+#    @enabled.setter
+#    def enabled(self, values):
+#        to_set = []
+#        for value in values:
+#            if value not in self.values:
+#                raise TypeError('%r is an invalid value. Choose one of %r.' % (value, self.values))
+#            to_set.append(value)
+#        self._enabled_values = to_set
+
+
+class AMTKVM(wsmanModule.wsmanModule):
+    '''Control over a device's KVM (VNC) functionality.'''
+    _RESOURCES = {
+        'kvmRedirectionSap': 'CIM_KVMRedirectionSAP',
+        'kvmRedirectionSettingData': 'IPS_KVMRedirectionSettingData',
+        'redirectionService': 'AMT_RedirectionService',
+        'tlsSettingData': 'AMT_TLSSettingData',
+    }
+
+    @property
+    def enabled(self):
+        '''
+        Whether KVM functionality is enabled or disabled.
+
+        True/False
+
+        .. note:: This will return True even if KVM is enabled, but no ports for it
+           are.
+        '''
+        e_state = self.RESOURCES['kvmRedirectionSap'].get('EnabledState')
+        return AMT_KVM_ENABLEMENT_MAP[e_state][0]
+
+    @enabled.setter
+    def enabled(self, value):
+        if value is True:
+            requested_state = 2
+        elif value is False:
+            requested_state = 3
+        else:
+            raise TypeError('Please specify Either True or False.')
+        self.RESOURCES['kvmRedirectionSap'].invoke(
+            'RequestStateChange',
+            state = requested_state
+        )
+
+    @property
+    def enabled_ports(self):
+        '''Tells you (and/or allows you to set) the enabled ports for VNC.'''
+        ports = set()
+        if self.RESOURCES['kvmRedirectionSettingData'].get('Is5900PortEnabled'):
+            ports.add(5900)
+        if self.RESOURCES['redirectionService'].get('ListenerEnabled'):
+            ports.add(16994)
+            if self.RESOURCES['tlsSettingData'].enumerate()['AMT_TLSSettingData'][0]['Enabled']:
+                ports.add(16995)
+        return ports
+#
+#    @enabled_ports.setter
```

Comment:
Is this chunk supposed to be here? Or can it be deleted?

Reply:
I've deleted it